### ISSUE TITLE
Clang

### ISSUE THREAD

**Issue by ajneu** [arobenko/comms_champion#1]
Just out of interest, I've tried compiling this codebase with Clang.

There are some interesting diagnostics, that you might want to check out.

Using clang:
```bash
sudo apt-get install clang                     # clang
export CC=clang; export CXX=clang++

mkdir build
cd    build
cmake ..
```

To get clang to work, add another `elseif` [here](https://github.com/arobenko/comms_champion/blob/320fd23debaff663e8355e6e702114c52b365404/CMakeLists.txt#L90) with something along these lines:

```cmake
elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
    set (extra_flags_list
        "--std=c++11"
        # "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wctor-dtor-privacy"
        # "-Wmissing-include-dirs"
        # "-Woverloaded-virtual" "-Wredundant-decls" "-Wshadow" "-Wundef" "-Wunused"
        # "-Wno-unknown-pragmas" "-fdiagnostics-show-option"

	## "-Wnoexcept" "-Wlogical-op" "-Wstrict-null-sentinel" # not supported under clang
    )
    
    if (NOT CC_NO_WARN_AS_ERR)
        list (APPEND extra_flags_list "-Werror")
    endif ()
    
    string(REPLACE ";" " " extra_flags "${extra_flags_list}")
    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${extra_flags}")
    
    if (CC_STATIC_RUNTIME)
        SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} -static-libstdc++ -static-libgcc")
    endif ()
```

As you can see, I've commented out a lot of warnings above, which can be put back later...

Even without these warnings, clang detects a lot of things... such as...

*  `comms/include/comms/util/alloc.h` line 283
```diff
-        return iter->alloc<TObj>(std::forward<TArgs>(args)...);
+        return iter->template alloc<TObj>(std::forward<TArgs>(args)...);
```

* `comms/test/CommsTestCommon.h` line 44
```diff
-    using Base::Message;
+    using typename Base::Message;
```

etc.
You'll notice that Clang is quite strict (or "restrictive"). ([Example](https://bugs.llvm.org/show_bug.cgi?id=31976) regarding [this line in embxx](https://github.com/arobenko/embxx/pull/8/files#diff-632c1b24566a91b5c21a2a771b79c584L184), ...)
Perhaps some things are somewhat more manageable, if c++14 is allowed...

**Comment by arobenko** [comment]
Hi Albert,
Thanks for your work with clang support, here and in embxx project. At this moment I'm a bit busy implementing support for ublox-8 receivers in my [ublox](https://github.com/arobenko/ublox) project. Supporting clang is the next task in my TODO list. 

These are all my personal projects. I'm spending around 6-8 hours a week on them during my commute time to work and back, so the progress is a bit slow. My current estimation is that I'll be able to look at clang support in April - May this year.

**Comment by arobenko** [comment]
Hi Albert,
I finally came around to look at clang support for this project. 
After fixing obvious and easy errors of adding "template" or "typename" keywords I discovered, the clang doesn't like valid C++11 constructs such as:
```
auto func() -> decltype(_ref_returning_expression_)
{
    return _ref_returning_expression_;
}
```
However, it doesn't have any problem with valid following c++14 expression:
```
auto& func() 
{
    return _ref_returning_expression_;
}
```
I had to modify my code to use second version when C++14 standard is requested.

At this moment "develop" branch supports compilation with clang compiler, but only if CMAKE_CXX_STANDARD variable is set to 14. It will find its way to master branch soon as well.

Thanks for your testing effort and guidance.
